### ISSUE TITLE
"Java heap space" error thrown exporting a large csv file

### ISSUE THREAD

**Issue by infomofo** [metabase/metabase#5187]
Thanks for being part of the Metabase project!
-------
- Mac OS 10.12.3
- Chrome 58.0.3029
- Metabase v0.24.1
- Redshift 
- Elastic Beanstalk

Exporting a large file breaks metabase and shows a json error in the browser.

```
{
message: "Java heap space",
stacktrace: [
"api.dataset$export_to_csv.invokeStatic(dataset.clj:55)",
"api.dataset$export_to_csv.invoke(dataset.clj:52)",
"api.dataset$as_format.invokeStatic(dataset.clj:102)",
"api.dataset$as_format.invoke(dataset.clj:94)",
"api.dataset$fn__28513$fn__28514.invoke(dataset.clj:123)",
"api.common.internal$do_with_caught_api_exceptions.invokeStatic(internal.clj:229)",
"api.common.internal$do_with_caught_api_exceptions.invoke(internal.clj:224)",
"api.dataset$fn__28513.invokeStatic(dataset.clj:116)",
"api.dataset$fn__28513.invoke(dataset.clj:116)"
]
}
```
![image](https://cloud.githubusercontent.com/assets/539599/26427980/41e0d61c-40ad-11e7-9d7a-8953f59fecb0.png)

**Comment by salsakran** [comment]
we should catch this and display a friendlier error page.

**Comment by jamesbloomer** [comment]
Is there a particular reason this error happens? We're trying to return a table of of ~200K rows with a few joins and are seeing this error.

**Comment by camsaul** [comment]
It happens because the JVM doesn't have enough memory to process the request. What size instance are you using on Elastic Beanstalk?

**Comment by jamesbloomer** [comment]
We're running on a t2.medium. After some testing it appears that the CSV works fine but the XLSX export fails, I guess that's creating more objects.

**Comment by demorenoc** [comment]
I just run into the same error for a ~34K row x 20 columns query when downloading XLSX. I am also running in AWS EB with a t2.medium.

**Comment by thibault-rouby** [comment]
When exporting a query on XLSX, my metabase instance memory skyrocket until the Java heap space memory error is thrown.
When exporting a query on CSV, it works fine.
The output file is less then 200MB but with the XLSX tentative export, the memory usage goes over 15GB...
It's running on docker, on a stand-alone server

**Comment by thucnc** [comment]
Hello, I got the same issue when exporting (20k rows * 11 columns) raw data to csv/xlsx on metabase v0.25.1 docker container. 

In addition, when exporting smaller data set, say 5k or 1k rows, I observed RAM usage increased, but not release afterwards. (Redash released memory with the same test case). This situation makes the RAM usage gradually increase and reach the memory limit at late of work-day. 

Rgds,

**Comment by thucnc** [comment]
@salsakran may you give this bug higher priority as it slows down our team productivity a lot? Anyway, Metabase is really outstanding thanks for your team.

**Comment by marcelmfs** [comment]
@thucnc metabase's relying on docjure to export xlsx, and docjure still don't have stream output writer implemented yet, as it's a somewhat low activity project (last commit as of Oct/2016).

**Comment by thucnc** [comment]
Hello @salsakran, the docjure library causing this issue hasn't been fixed yet, so we are still waiting for it ?

**Comment by hai-ld** [comment]
Is there any workaround for this issue? Can we limit the number of rows in one export? Or can we selectively disable XLSX export?

**Comment by thucnc** [comment]
We are still suffering from this issue  (many non-tech staff still need to download results) so this bug is really annoying. Is there any workaround ?

**Comment by jornh** [comment]
@thucnc, @hai-ld looks like the PR https://github.com/mjul/docjure/pull/47 with addition of streaming support got stuck on the lack of missing test code demonstrating that it supports existing features of the Docture library. I think that’s a fair request from the Docture maintainer.

Until that fix can happen - which in turn will hopefully also better .xlsx export memory consumption with Metabase here a few work-around suggestions:
- Buy more memory
- Teach users to apply (stricter) filters - or reduce number of columns - or export as CSV
- By looking at your business processes eliminate or reduce the need for .xlsx exports. It is IMO a sign of suboptimal processes if you have to manually move data. Maybe you can bring the Excel consumers to Metabase. Maybe you can automate import of data in another system if that is what the Excel files are used for.

**Comment by adamantines3** [comment]
Getting the same error while exporting as a csv file.I can successfully export a table 2x the size but not this one, it has 50 million rows and 5 columns.

**Comment by fiendish** [comment]
I'm getting this when trying to export an excel file for a mere 5991 rows by 13 columns. The 1.7MiB csv download converted to xlsx is 891KiB. How much memory could metabase possibly need for that? 😭

**Comment by camsaul** [comment]
This was fixed by #11832
